File: source/python/thread_safe_queue.py
import threading
import queue

q = queue.Queue(maxsize=100)
num_producers = 3
items_per_producer = 5
total_items = num_producers * items_per_producer
results = []
results_lock = threading.Lock()

def producer(prod_id):
    for i in range(items_per_producer):
        q.put(f"p{prod_id}-{i}")

def consumer():
    received = []
    for _ in range(total_items):
        val = q.get()
        received.append(val)
    with results_lock:
        results.extend(received)

consumer_thread = threading.Thread(target=consumer)
producer_threads = [threading.Thread(target=producer, args=(i,)) for i in range(num_producers)]

consumer_thread.start()
for t in producer_threads:
    t.start()

for t in producer_threads:
    t.join()
consumer_thread.join()

print(f"consumed {len(results)} items")
results.sort()
print(f"values: {' '.join(results)}")

File: source/rust/thread_safe_queue.rs
use std::sync::{Arc, Mutex, mpsc};
use std::thread;

fn main() {
    let num_producers = 3;
    let items_per_producer = 5;
    let total_items = num_producers * items_per_producer;

    let (tx, rx) = mpsc::channel::<String>();
    let results = Arc::new(Mutex::new(Vec::new()));
    let results_c = Arc::clone(&results);

    // Start consumer
    let consumer = thread::spawn(move || {
        let mut received = Vec::new();
        for _ in 0..total_items {
            received.push(rx.recv().unwrap());
        }
        *results_c.lock().unwrap() = received;
    });

    // Start producers
    let mut producers = Vec::new();
    for id in 0..num_producers {
        let tx = tx.clone();
        producers.push(thread::spawn(move || {
            for i in 0..items_per_producer {
                tx.send(format!("p{}-{}", id, i)).unwrap();
            }
        }));
    }

    for p in producers { p.join().unwrap(); }
    consumer.join().unwrap();

    let mut res = results.lock().unwrap().clone();
    println!("consumed {} items", res.len());
    res.sort();
    println!("values: {}", res.join(" "));
}
